shop4All-2 Log in and register

Osnaga-Robert opened the pull request:
- delete .formLogin from WebSecurityConfiguration
- moved JwtRequest and JwtResponse to DTO package
- all controllers return ResponseEntity
- resolved comments order
- deleted templates properties
- created Factory Pattern for seller, buyer and admin

Osnaga-Robert committed:
Updates

Update log in and register

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/Controller/UserController.java:
```diff
@@ -19,17 +19,32 @@ public class UserController {
     @Autowired
     private UserService userService;
 
+
+
     //initialize a demo admin
     @PostConstruct
     public void initRolesAndUsers(){
         userService.initRolesandUser();
     }
```
Delete line 22 and 23 since we don't need so much space in our code.

Be careful at indentation.

Structure:

public void initRolesAndUsers() {

This way you ensure readability of the code and avoid branch conflicts caused by bad indentation of the code. Try to indent all code using IDE plugins or functionalities.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/Controller/UserController.java:
```diff
@@ -19,17 +19,32 @@ public class UserController {
+    // Handle POST requests to /registerNewSeller to register a new seller
     @PostMapping({"/registerNewSeller"})
-    public ResponseEntity<?> registerNewSeller(@RequestBody User user) {
+    public ResponseEntity<?> registerNewSeller(@RequestBody Seller seller) {
```
If you know the type of the returned ResponseEntity, please add the Object type between "<>". Do the same for every ResponseEntity<?>.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/Entity/UserFactory/Admin.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Set;
+
+@Entity
+public class Admin implements User{
```
Add a space after User ( User { ).

PatrickDiallo23 commented on shop4All-backend/target/classes/application.properties:
The target folder is generated when you build the backend application so it is not needed to push the target folder.

PatrickDiallo23 reviewed (changes requested):
Take a look at the comments. I will leave in this comment the other suggestions/questions:

- Don't push the node_modules folder because you are increasing the repository size and this folders is initialized when you install all the dependencies using npm install. The role of node_modules folder is just to get the npm dependencies.
- Was it needed to move all the entities in the UserFactory package? The entities objects should be in the "entity" or "entities" package.
- Make sure that you changed OAuth2 credentials since they appear in the commit history.
- Is there any way to avoid initializing 3 types of different repos in the UserService and use a single object that does all the stuff for you? This way you can avoid the redundant use of if else blocks. Or you can extract the logic where you used if else blocks in a separate method.
- In the database diagram you mentioned that you will use Integer id, not a String one. 
- You can create your own custom exceptions (in exceptions package) to help you manage the errors in a more efficient way.
- The packages should be named using lowercase letters.
- Make sure that all code is indented.

Osnaga-Robert committed:
Updates

Update log in and register

Osnaga-Robert commented:
- created UserService and UserEntity
- removed AdminService, SellerService, BuyerService and their entities
- created custom exceptions(only for basic exceptions)
- Integed Id added
- register validate email and password
- changed the name of the packages, but omitted to put them in the plural

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/controller/JwtController.java:
In the future, you can use the logic for creating the JWT Token in the UserController class.

PatrickDiallo23 commented on shop4All-backend/target/classes/com/example/shop4All_backend/Configuration/WebSecurityConfiguration.class:
Try to get rid of the target folder entirely.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/service/JwtService.java:
```diff
@@ -79,48 +64,37 @@ else if (seller.isPresent()) {
-            throw new Exception("User is disable");
-        } catch (BadCredentialsException e){
-            throw new Exception("Bad credential for user");
+            throw new LogInException("User is disabled");
+        } catch (BadCredentialsException e) {
+            throw new LogInException("Bad credentials for user");
+        } catch (AuthenticationException e) {
+            throw new LogInException("Authentication failed");
```
Did you tested if all the exceptions are called if something went wrong? I am asking because you have to make sure that the order of the Exceptions classes is okay so that all the catch block will be executed if you get an exception, not just one block.

For example, you should put the Exception catch block last because every other exceptions are derived from Exception.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/service/JwtService.java:
```diff
@@ -79,48 +64,37 @@ else if (seller.isPresent()) {
+            throw new LogInException("Authentication failed");
         }
     }
+
+
```
Those 2 spaces are not useful here. You can delete them.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/entity/User.java:
You can add @Setter and @Getter at the level of the class.

PatrickDiallo23 reviewed (changes requested):
Please take a look at the newest comments.

Osnaga-Robert committed:
Updates

Log in, register and basic exception handler

Osnaga-Robert commented:
- Resolved basic exception handler
- Added @Setter and @Getter
- Deleted target folder

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/services/JwtService.java:
```diff
@@ -49,11 +47,12 @@ public JwtResponse createJwtToken(JwtRequest jwtRequest) throws AuthenticationEx
             user = userOptional.get();
         }
         else{
+            System.out.println("---------");
```
In a backend application, you will be using log (Logger) instead of prints. Take this comment into consideration in the future PR. You can also refer to this [link](https://www.baeldung.com/java-system-out-println-vs-loggers#:~:text=With%20the%20use%20of%20Loggers,timestamp%2C%20method%20name%2C%20etc.).

PatrickDiallo23 reviewed (approved):
Looks good to me. I have added an additional comment for the future PRs to be taken into consideration. Also, be careful that, when you will be implementing the frontend side, some logic in your code (for example initRoleandUser/[initRoleAndUser](https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html) method) should be done on the frontend side.
